File: rvc_service/RVCSDK/examples/Python/GetExternalColorCameraPointMap.py
import PyRVC as RVC
import numpy as np
import cv2
import os
from Utils.Tools import *

# This example shows the process of acquiring the point cloud of an external camera. Note that this program cannot be
# run directly. The part that acquires the image of the external camera needs to be modified.
if __name__ == "__main__":
    # Prepare work: Users need to calibrate the external camera themselves. 
    # The quality of calibration directly affects the quality of final result.
    external_camera_intrinsic_matrix = [9609.9961, 0, 1231.808, 0, 9608.9238, 1020.5879, 0, 0, 1]    # Needs to be modified.
    external_camera_camera_distortion = [0.21287285, -1.2765415, 0.00071774476, -0.0017654195, -11.771881]  # Needs to be modified.
    external_camera_image_width = 2448;                                             # Needs to be modified.
    external_camera_image_height = 2048;                                            # Needs to be modified.
    
    # 1. Initialize
    RVC.SystemInit()
    opt = RVC.SystemListDeviceTypeEnum.All
    ret, devices = RVC.SystemListDevices(opt)
    if len(devices) == 0:
        print("Can not find any RVC Camera!")
        RVC.SystemShutdown()
        exit(1)
    if devices[0].IsFirmwareMatch() == False:
        print("device firmware mismatch, Please use RVCManager to upgrade the firmware")
        RVC.SystemShutdown()
        exit(1)
    x = RVC.X2.Create(devices[0])
    ret1 = x.Open()
    if x.IsOpen() == False:
        print("Failed to open camera! Please check whether the camera is connected and make sure it is not occupied and supports X2.")
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        exit(1)
    ret, info = devices[0].GetDeviceInfo()
    if info.support_extra:
        print("Device has already support color camera!")
        RVC.SystemShutdown()
        exit(1)

    save_dir = "Data"
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
        
    # 2. Capture three point maps with 4*11 calibration board to get external camera extrinsic matrix.
    # 2.1 capture first point map
    input("Put the calibration board at the nearest working distance and take first frame. Press Enter to continue...")
    ret = x.Capture()
    internal_camera_point_map0 = x.GetPointMap().Clone()        
    internal_camera_image0 = x.GetImage(RVC.CameraID_Left).Clone()
    internal_camera_point_map0.SaveWithImage("{}/calibration_internal0.ply".format(save_dir), internal_camera_image0, RVC.PointMapUnitEnum.Meter, True)
    internal_camera_image0.SaveImage(save_dir + "/calibration_internal0.png")

    external_camera_image0 = np.array([...])  # External image should be numpy array. Needs to be modified.

    # 2.2 capture second point map
    input("Put the calibration board at the middle working distance and take second frame. Press Enter to continue...")
    ret = x.Capture()
    internal_camera_point_map1 = x.GetPointMap().Clone()
    internal_camera_image1 = x.GetImage(RVC.CameraID_Left).Clone()
    internal_camera_point_map1.SaveWithImage("{}/calibration_internal1.ply".format(save_dir), internal_camera_image1, RVC.PointMapUnitEnum.Meter, True)
    internal_camera_image1.SaveImage(save_dir + "/calibration_internal1.png")
    external_camera_image1 = np.array([...])  # External image should be numpy array. Needs to be modified.

    # 2.3 capture third point map
    input("Put the calibration board at the furthest working distance and take third frame. Press Enter to continue...")
    ret = x.Capture()
    internal_camera_point_map2 = x.GetPointMap().Clone()
    internal_camera_image2 = x.GetImage(RVC.CameraID_Left).Clone()
    internal_camera_point_map2.SaveWithImage("{}/calibration_internal2.ply".format(save_dir), internal_camera_image2, RVC.PointMapUnitEnum.Meter, True)
    internal_camera_image2.SaveImage(save_dir + "/calibration_internal2.png")
    external_camera_image2 = np.array([...])  # External image should be numpy array. Needs to be modified.

    # 2.4 get external camera extrinsic matrix
    error_code, external_camera_extrinsic_matrix, error = RVC.GetExternalCameraExtrinsicMatrix(
            internal_camera_image0, internal_camera_point_map0, external_camera_image0, 
            internal_camera_image1, internal_camera_point_map1, external_camera_image1, 
            internal_camera_image2, internal_camera_point_map2, external_camera_image2, 
            external_camera_image_width, external_camera_image_height,
            external_camera_intrinsic_matrix,
            external_camera_camera_distortion)
    if error_code != 0:
        print("Get external camera extrinsic matrix failed! Error code: ", error_code)
        x.Close()
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        exit(0)

    print("External camera extrinsic matrix: ", external_camera_extrinsic_matrix)
    print("Reprojection error: ", error)
    
    # 3. After calibration, we can capture a point map with RVC camera and the corresponding image by external camera and get external camera point map.
    ret = x.Capture()
    internal_camera_point_map = x.GetPointMap()        
    internal_camera_image = x.GetImage(RVC.CameraID_Left)
    internal_camera_point_map.SaveWithImage("{}/scene_internal.ply".format(save_dir), internal_camera_image, RVC.PointMapUnitEnum.Meter, True)
    internal_camera_image.SaveImage(save_dir + "/scene_internal.png")
    external_camera_point_map = RVC.GetExternalCameraPointMap(
        internal_camera_point_map, 
        external_camera_image_width, external_camera_image_height, 
        external_camera_intrinsic_matrix, external_camera_camera_distortion, external_camera_extrinsic_matrix)
    external_camera_point_map.Save("{}/scene_external.ply".format(save_dir), RVC.PointMapUnitEnum.Meter, True)
            
    x.Close()
    RVC.X2.Destroy(x)
    RVC.SystemShutdown()

File: rvc_service/RVCSDK/examples/CSharp/GetExternalColorCameraPointMap.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;
using static RVC_CSharp.ExternalColorCamera;

namespace RVC_CSharp
{
    class Program
    {
        // This example shows the process of acquiring the point cloud of an external camera. Note that this program cannot be
        // run directly. The part that acquires the image of the external camera needs to be modified.
        static void Main()
        {
            #region Step 0 ,Init & Find & Open

            if (false == RVC_CSharp.System.Init())
            {
                Console.WriteLine("Failed to init system.");
                Console.ReadKey();
                return;
            }

            Device device = RVC_CSharp.System.FindDeviceByIndex(0);
            if (false == device.IsValid() || false == device.CheckCanConnected())
            {
                Console.WriteLine("Device Can not be Connected.\nPlease Check Connection.");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            if(device.IsFirmwareMatch() == false)
            {
                Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            X2 camera = RVC_CSharp.X2.Create(device);
            camera.Open();
            if (false == camera.IsValid() || false == camera.IsOpen())
            {
                Console.WriteLine($"Failed to open camera .Error: {System.GetLastErrorMessage()} ");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            DeviceInfo info = new DeviceInfo();
            device.GetDeviceInfo(ref info);
            if (info.support_extra == true)
            {
                Console.WriteLine("Device has already support color!");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            #endregion

            #region Step 1 ,Get External Camera Extrinsic Matrix
            // Prepare work: Users need to calibrate the external camera themselves. 
            // The quality of calibration directly affects the quality of final result.
            float[] external_camera_intrinsic_matrix = { 2400, 0, 1000, 0, 2400, 750, 0, 0, 1 }; ;         // Needs to be modified.
            float[] external_camera_camera_distortion = { -0.099f, 0.218f, 0.00037f, 0.0001f, -0.26f };  // Needs to be modified.
            int external_camera_image_width = 0;                                                        // Needs to be modified.
            int external_camera_image_height = 0;                                                       // Needs to be modified. 

            // Capture three point maps with 4 * 11 calibration board to get external camera extrinsic matrix.
            float[] external_camera_extrinsic_matrix = new float[16];

            bool ret = false;

            unsafe
            {
                Directory.CreateDirectory($"./{info.name}-{info.sn}");
                // Capture First Point Map
                Console.WriteLine("Put the calibration board at the nearest working distance and take first frame.");
                Console.WriteLine("Press Enter to continue...");

                Console.ReadLine();

                ret = camera.Capture();
                if (!ret)
                {
                    Console.WriteLine($"Failed to capture.Error: {System.GetLastErrorMessage()} ");
                    camera.Close();
                    camera.Destroy();
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                    return;
                }

                PointMap internalCameraPointMap0 = camera.GetPointMap();

                Image internalCameraImage0 = camera.GetImage(CameraID.CameraID_Left);

                byte[] externalCameraImageData0 = null;// Get external image. Needs to be modified.

                string pmAddr = $"./{info.name}-{info.sn}/calibration_internal0.ply";
                string imgAddr = $"./{info.name}-{info.sn}/calibration_internal0.ply";
                internalCameraPointMap0.SaveColorPointCloud(internalCameraImage0, pmAddr, unit_mm: false);
                internalCameraImage0.SaveImage(imgAddr);

                // Capture Second Point Map
                Console.WriteLine("Put the calibration board at the middle working distance and take first frame.");
                Console.WriteLine("Press Enter to continue...");
                Console.ReadLine();

                ret = camera.Capture();
                if (!ret)
                {
                    Console.WriteLine($"Failed to capture.Error: {System.GetLastErrorMessage()} ");
                    camera.Close();
                    camera.Destroy();
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                    return;
                }

                PointMap internalCameraPointMap1 = camera.GetPointMap();

                Image internalCameraImage1 = camera.GetImage(CameraID.CameraID_Left);

                byte[] externalCameraImageData1 = null;// Get external image. Needs to be modified.

                pmAddr = $"./{info.name}-{info.sn}/calibration_internal1.ply";
                imgAddr = $"./{info.name}-{info.sn}/calibration_internal1.ply";
                internalCameraPointMap1.SaveColorPointCloud(internalCameraImage1, pmAddr, unit_mm: false);
                internalCameraImage1.SaveImage(imgAddr);

                // Capture Second Point Map
                Console.WriteLine("Put the calibration board at the furthest working distance and take first frame.");
                Console.WriteLine("Press Enter to continue...");
                Console.ReadLine();

                ret = camera.Capture();
                if (!ret)
                {
                    Console.WriteLine($"Failed to capture.Error: {System.GetLastErrorMessage()} ");
                    camera.Close();
                    camera.Destroy();
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                    return;
                }

                PointMap internalCameraPointMap2 = camera.GetPointMap();

                Image internalCameraImage2 = camera.GetImage(CameraID.CameraID_Left);

                byte[] externalCameraImageData2 = null;// Get external image. Needs to be modified.

                pmAddr = $"./{info.name}-{info.sn}/calibration_internal1.ply";
                imgAddr = $"./{info.name}-{info.sn}/calibration_internal1.ply";
                internalCameraPointMap2.SaveColorPointCloud(internalCameraImage2, pmAddr, unit_mm: false);
                internalCameraImage2.SaveImage(imgAddr);

                // Get External Camera Extrinsic Matrix
                double error;
                int errorCode = GetExternalCameraExtrinsicMatrix(internalCameraImage0.m_handle,
                    internalCameraPointMap0.m_handle,
                    externalCameraImageData0, internalCameraImage1.m_handle, internalCameraPointMap1.m_handle,
                    externalCameraImageData1, internalCameraImage2.m_handle, internalCameraPointMap2.m_handle,
                    externalCameraImageData2, external_camera_image_width, external_camera_image_height,
                    external_camera_intrinsic_matrix, external_camera_camera_distortion,
                    external_camera_extrinsic_matrix, out error);

                if (errorCode != 0)
                {
                    Console.WriteLine($"Get external camera extrinsic matrix failed! Error: {System.GetLastErrorMessage()} ");
                    camera.Close();
                    camera.Destroy();
                    RVC_CSharp.System.Shutdown();
                    Console.WriteLine("Press any key to exit. ");
                    Console.ReadKey();
                    return;
                }

                // print extrinsic matrix and error
                Console.WriteLine("External camera extrinsic matrix:");
                Console.WriteLine(string.Join(", ", external_camera_extrinsic_matrix));
                Console.WriteLine($"Reprojection error(pixel): {error}");
            }
            // Capture three point maps with 4*11 calibration board to get external camera extrinsic matrix.

            #endregion

            #region Step 2 , Capture With External Color Camera

            ret = camera.Capture();

            if (false == ret)
            {
                Console.WriteLine($"Failed to Capture .Error: {System.GetLastErrorMessage()} ");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            Image image = camera.GetImage(CameraID.CameraID_Left);
            DepthMap depth = camera.GetDepthMap();
            PointMap pointMap = camera.GetPointMap();

            // 2D Image
            string imageFile = $"./{info.name}-{info.sn}/scene_internal.png";
            Console.WriteLine($"Save Image,path = {imageFile}");
            image.SaveImage(imageFile);

            // Point Cloud 
            string pointCloudFile = $"./{info.name}-{info.sn}/scene_internal.ply";
            Console.WriteLine($"Save PointMap,path = {pointCloudFile}");
            pointMap.SaveColorPointCloud(image, pointCloudFile, false);

            PointMap external_camera_point_map = GetExternalCameraPointMap(pointMap.m_handle, external_camera_image_width, external_camera_image_height,
        external_camera_intrinsic_matrix, external_camera_camera_distortion, external_camera_extrinsic_matrix);

            string pointCloudFile2 = $"./{info.name}-{info.sn}/scene_internal.ply";
            external_camera_point_map.Save(pointCloudFile2, PointMapUnit.Meter, true);

            camera.Close();
            camera.Destroy();

            RVC_CSharp.System.Shutdown();
            Console.WriteLine("System close.");

            Console.ReadKey();
            #endregion

            return;
        }

    }
}
